source/fuzz/transformation_duplicate_region_with_selection.cpp

```diff
@@ -0,0 +1,566 @@
+  message_.set_exit_block_id(exit_block_id);
+  *message_.mutable_original_label_to_duplicate_label() =
+      fuzzerutil::MapToRepeatedUInt32Pair(original_label_to_duplicate_label);
+  *message_.mutable_original_id_to_duplicate_id() =
+      fuzzerutil::MapToRepeatedUInt32Pair(original_id_to_duplicate_id);
+  *message_.mutable_original_id_to_phi_id() =
+      fuzzerutil::MapToRepeatedUInt32Pair(original_id_to_phi_id);
+}
+
+bool TransformationDuplicateRegionWithSelection::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // Instruction with the id |condition_id| must exist and must be of a bool
+  // type.
+  auto bool_instr =
+      ir_context->get_def_use_mgr()->GetDef(message_.condition_id());
+  if (bool_instr == nullptr || !bool_instr->type_id()) {
+    return false;
+  }
+  if (!ir_context->get_type_mgr()->GetType(bool_instr->type_id())->AsBool()) {
+    return false;
+  }
+
+  // The |new_entry_fresh_id| must be fresh and distinct.
+  std::set<uint32_t> ids_used_by_this_transformation;
+  if (!CheckIdIsFreshAndNotUsedByThisTransformation(
+          message_.new_entry_fresh_id(), ir_context,
+          &ids_used_by_this_transformation)) {
+    return false;
+  }
+
+  // The |merge_label_fresh_id| must be fresh and distinct.
+  if (!CheckIdIsFreshAndNotUsedByThisTransformation(
+          message_.merge_label_fresh_id(), ir_context,
+          &ids_used_by_this_transformation)) {
+    return false;
+  }
+
+  // The entry and exit block ids must refer to blocks.
+  for (auto block_id : {message_.entry_block_id(), message_.exit_block_id()}) {
+    auto block_label = ir_context->get_def_use_mgr()->GetDef(block_id);
+    if (!block_label || block_label->opcode() != SpvOpLabel) {
+      return false;
+    }
+  }
+  auto entry_block = ir_context->cfg()->block(message_.entry_block_id());
+  auto exit_block = ir_context->cfg()->block(message_.exit_block_id());
+
+  // The |entry_block| and the |exit_block| must be in the same function.
+  if (entry_block->GetParent() != exit_block->GetParent()) {
+    return false;
+  }
+
+  // The |entry_block| must dominate the |exit_block|.
+  auto dominator_analysis =
+      ir_context->GetDominatorAnalysis(entry_block->GetParent());
+  if (!dominator_analysis->Dominates(entry_block, exit_block)) {
+    return false;
+  }
+
+  // The |exit_block| must post-dominate the |entry_block|.
+  auto postdominator_analysis =
+      ir_context->GetPostDominatorAnalysis(entry_block->GetParent());
+  if (!postdominator_analysis->Dominates(exit_block, entry_block)) {
+    return false;
+  }
+
+  auto enclosing_function = entry_block->GetParent();
+
+  // |entry_block| cannot be the first block of the |enclosing_function|.
+  if (&*enclosing_function->begin() == entry_block) {
+    return false;
+  }
+
+  auto region_set = GetRegionBlocks(ir_context, entry_block, exit_block);
+
+  // Check whether |region_set| really is a single-entry single-exit region, and
```

Comment:
Can you open an issue to document the fact that this duplicates code in TransformationOutlineFunction, and that we should consider a refactoring, and then insert a TODO(...) comment here?